### PR TITLE
Issue #557: refactoring models

### LINKED ISSUE #557: Refactoring models
## Subject of the issue

We are now building models via functions. I'd like to refactor it into classes so that we can use the inheritance between classes to simplify part of the code. I personally find the current code quite confusing.

This is also a preparation for supporting more loss possibilities. However, there might be a risk that some demo will no longer work.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Rewrite registration models from functions to classes, and use registry.
Moreover, affine model is merged into DDF model and got test covered (it was never tested I believe)

Fixes #557

Demos need to be updated
- [x] grouped_mask_prostate_longitudinal, updated
- [x] grouped_mr_heart, updated
- [x] paired_ct_lung, DGX tmux 0, training, ETA 9am Monday
- [x] paired_mr_brain, will not be updated as the ckpt meant to be improved further, see #577.
- [x] paired_mrus_prostate, updated, but need further attention #621
- [x] unpaired_ct_abdomen, will not be updated as the ckpt meant to be improved further, see #552.
- [x] unpaired_ct_lung, PT5
- [x] unpaired_mr_brain, will not be updated as the ckpt meant to be improved further, see #620.
- [x] unpaired_us_prostate_cv, DGX tmux

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
ooops, as I changed the network definitions, we may not be able to load pretrained weights :(

Then it is important to have some automated scripts to retrain the models one by one.

**Comment by mathpluscode** [comment]
@YipengHu @s-sd @zacbaum This PR is ready for review.

Except that I'm retraining all models to regenerate the checkpoints, as change of network architectures makes all ckpts not suitable. Therefore it would be nice/saving time if you review this starting now ^^.

**Comment by mathpluscode** [comment]
@YipengHu @s-sd @zacbaum This PR is ready for review now! XD

**Review by s-sd — APPROVED** [review]
@mathpluscode looks good to me. If you need help updating any of the demos before merging, let me know. Also, not sure if required but if you need re-approval after updating some more demos, let me know as well.

**Comment by mathpluscode** [comment]
> @mathpluscode looks good to me. If you need help updating any of the demos before merging, let me know. Also, not sure if required but if you need re-approval after updating some more demos, let me know as well.

Oh would I need to update some figures?

**Comment by s-sd** [comment]
> > @mathpluscode looks good to me. If you need help updating any of the demos before merging, let me know. Also, not sure if required but if you need re-approval after updating some more demos, let me know as well.
> 
> Oh would I need to update some figures? @s-sd

Yeah, I think would be a good idea to update them. I have a script written which creates and saves all figures so if you want I'd be happy to update figures once you're done updating the demo checkpoints, just let me know once done.

**Comment by mathpluscode** [comment]
> > > @mathpluscode looks good to me. If you need help updating any of the demos before merging, let me know. Also, not sure if required but if you need re-approval after updating some more demos, let me know as well.
> > 
> > 
> > Oh would I need to update some figures? @s-sd
> 
> Yeah, I think would be a good idea to update them. I have a script written which creates and saves all figures so if you want I'd be happy to update figures once you're done updating the demo checkpoints, just let me know once done.

@s-sd is your script uploaded in the repo?

**Comment by s-sd** [comment]
> > > > @mathpluscode looks good to me. If you need help updating any of the demos before merging, let me know. Also, not sure if required but if you need re-approval after updating some more demos, let me know as well.
> > > 
> > > 
> > > Oh would I need to update some figures? @s-sd
> > 
> > 
> > Yeah, I think would be a good idea to update them. I have a script written which creates and saves all figures so if you want I'd be happy to update figures once you're done updating the demo checkpoints, just let me know once done.
> 
> @s-sd is your script uploaded in the repo?

No, I just have it locally. Should I add it to the repo?

**Comment by mathpluscode** [comment]
> > > > > @mathpluscode looks good to me. If you need help updating any of the demos before merging, let me know. Also, not sure if required but if you need re-approval after updating some more demos, let me know as well.
> > > > 
> > > > 
> > > > Oh would I need to update some figures? @s-sd
> > > 
> > > 
> > > Yeah, I think would be a good idea to update them. I have a script written which creates and saves all figures so if you want I'd be happy to update figures once you're done updating the demo checkpoints, just let me know once done.
> > 
> > 
> > @s-sd is your script uploaded in the repo?
> 
> @mathpluscode no, I just have it locally. Should I add it to the repo?

Maybe? we can leave that for a separate issue. Would you mind to send me the script via mail/teams.
I may need several modif as not all demos have been updated in the PR. (Because some demos are meant to be improved anyway afterwards and they do now show meaningful results for now.)

**Comment by mathpluscode** [comment]
I guess I just realised that this branch has no conflict with #559.

So it is ready to be merged before/after #559. As I didn't influence the model part.
@YipengHu @NMontanaBrown could you have a look ^^

File: demos/paired_mrus_brain/demo_train.py
Comment (YipengHu) [inline demos/paired_mrus_brain/demo_train.py:50]: what is the reason for changing this?

**Review by YipengHu — APPROVED** [review]

File: deepreg/model/loss/image.py
Comment (mathpluscode) [reply deepreg/model/loss/image.py:21]: Which reduction mode?

File: demos/paired_mrus_brain/demo_train.py
Comment (mathpluscode) [reply demos/paired_mrus_brain/demo_train.py:50]: Make it more robust and less like to cause errors. As taking all memory might make other users/programs unable to run code.

Also, it allows you to know easily if you are using the largest batch size you can use.

File: deepreg/model/loss/image.py
Comment (mathpluscode) [reply deepreg/model/loss/image.py:21]: We always do reduction to have a scalar loss. I change here to support multiple GPUs.

**Comment by mathpluscode** [comment]
@YipengHu are you happy with my replies? @NMontanaBrown do you have any further comments :)

File: demos/paired_mrus_brain/demo_train.py
Comment (YipengHu) [reply demos/paired_mrus_brain/demo_train.py:50]: i think you mentioned this before - to be fair...

File: deepreg/model/loss/image.py
Comment (YipengHu) [reply deepreg/model/loss/image.py:21]: fine with as it now, we need to careful when to perform the reduction - this really should be in one place outside/instead-of in the "loss function", to minimise the risk to have loss functions dependent on the batch size.

**Comment by YipengHu** [comment]
> @YipengHu are you happy with my replies? @NMontanaBrown do you have any further comments :)

all looks good to me.

**Comment by NMontanaBrown** [comment]
Let me check, one minute.

**Comment by NMontanaBrown** [comment]
Can we merge the config docs first? I am still waiting on a few comments/resolutions there. @YipengHu @mathpluscode

**Comment by mathpluscode** [comment]
> Can we merge the config docs first? I am still waiting on a few comments/resolutions there. @YipengHu @mathpluscode

There is no conflicts as this PR does not change config structure ^^.

**Comment by NMontanaBrown** [comment]
> > Can we merge the config docs first? I am still waiting on a few comments/resolutions there. @YipengHu @mathpluscode
> 
> There is no conflicts as this PR does not change config structure ^^. But yes, if you prefer this way. I can wait.

Actually, yes, because affine parameter is no longer passed right? @mathpluscode

**Comment by mathpluscode** [comment]
> > > Can we merge the config docs first? I am still waiting on a few comments/resolutions there. @YipengHu @mathpluscode
> > 
> > 
> > There is no conflicts as this PR does not change config structure ^^. But yes, if you prefer this way. I can wait.
> 
> Actually, yes, because affine parameter is no longer passed right? @mathpluscode

oooops, good point.

Affine model is integrated into DDF, so that DDF+GlobalNet = Affine, you are totally right!

**Comment by mathpluscode** [comment]
> > > Can we merge the config docs first? I am still waiting on a few comments/resolutions there. @YipengHu @mathpluscode
> > 
> > 
> > There is no conflicts as this PR does not change config structure ^^. But yes, if you prefer this way. I can wait.
> 
> Actually, yes, because affine parameter is no longer passed right? @mathpluscode

I've updated the config docs https://deepreg--602.org.readthedocs.build/en/602/docs/configuration.html, have a look? @NMontanaBrown

**Comment by mathpluscode** [comment]
I have to merge this now as user has experienced a multi-gpu bug, which got solved here.

**Pull request merged**
